x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/TransportGetProfilingAction.java

```diff
@@ -0,0 +1,343 @@
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.search.aggregations.bucket.terms.StringTerms;
+import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregationBuilder;
+import org.elasticsearch.search.aggregations.metrics.Sum;
+import org.elasticsearch.search.aggregations.metrics.SumAggregationBuilder;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.profiler.utils.MapExtractor;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+
+public class TransportGetProfilingAction extends HandledTransportAction<GetProfilingRequest, GetProfilingResponse> {
+    private final NodeClient nodeClient;
+    private final ThreadContext threadContext;
+
+    @Inject
+    public TransportGetProfilingAction(TransportService transportService, ActionFilters actionFilters, NodeClient nodeClient) {
+        super(GetProfilingAction.NAME, transportService, actionFilters, GetProfilingRequest::new);
+        this.nodeClient = nodeClient;
+        this.threadContext = transportService.getThreadPool().getThreadContext();
+    }
+
+    @Override
+    protected void doExecute(Task submitTask, GetProfilingRequest request, ActionListener<GetProfilingResponse> submitListener) {
+        try (var ignored = threadContext.newTraceContext()) {
+            EventsIndex mediumDownsampled = EventsIndex.MEDIUM_DOWNSAMPLED;
+            this.nodeClient.prepareSearch(mediumDownsampled.getName())
+                .setSize(0)
+                .setQuery(request.getQuery())
+                .setTrackTotalHits(true)
+                .execute(new ActionListener<>() {
+                    @Override
+                    public void onResponse(SearchResponse searchResponse) {
+                        long sampleCount = searchResponse.getHits().getTotalHits().value;
+                        EventsIndex resampledIndex = mediumDownsampled.getResampledIndex(request.getSampleSize(), sampleCount);
+                        searchEventGroupByStackTrace(request, resampledIndex, submitListener);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        submitListener.onFailure(e);
+                    }
+                });
+        }
+    }
+
+    private void searchEventGroupByStackTrace(
+        GetProfilingRequest request,
+        EventsIndex eventsIndex,
+        ActionListener<GetProfilingResponse> submitListener
+    ) {
+        GetProfilingResponseBuilder responseBuilder = new GetProfilingResponseBuilder();
+        this.nodeClient.prepareSearch(eventsIndex.getName())
+            .setTrackTotalHits(false)
+            .setQuery(request.getQuery())
+            .addAggregation(
```

Comment:
I think setting the parent task is should be done in all of the subsearch requests. I am thinking that it's not necessary to have a tree structure, but rather have all of the subsearches be the parent of the main profiling action? In that case you could use a ParentTaskAssigningClient?

Reply:
Yes, I think that should be sufficient. I'll use `ParentTaskAssigningClient` then.